Mathlib/CategoryTheory/Retract.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2024 Jack McKoen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jack McKoen
+-/
+import Mathlib.CategoryTheory.MorphismProperty.Basic
+import Mathlib.CategoryTheory.EpiMono
+
+/-!
+# Retracts
+
+Defines retracts of objects and morphisms.
+
+-/
+
+universe v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+/-- An object `X` is a retract of `Y` if there are morphisms `i : X ⟶ Y` and `r : Y ⟶ X` such
+that `i ≫ r = 𝟙 X`. -/
+structure Retract (X Y : C) where
+  /-- `i : X ⟶ Y` -/
+  i : X ⟶ Y
+  /-- `r : Y ⟶ X` -/
+  r : Y ⟶ X
+  /-- `i ≫ r = 𝟙 X` -/
+  retract : i ≫ r = 𝟙 X
+
+namespace Retract
+
+attribute [reassoc (attr := simp)] retract
+
+variable {X Y : C} (h : Retract X Y)
+
+/-- a retract determines a split epimorphism. -/
+@[simps] def splitEpi : SplitEpi h.r where
+  section_ := h.i
+
+/-- a retract determines a split monomorphism. -/
+@[simps] def splitMono : SplitMono h.i where
+  retraction := h.r
+
+instance : IsSplitEpi h.r := ⟨⟨h.splitEpi⟩⟩
+
+instance : IsSplitMono h.i := ⟨⟨h.splitMono⟩⟩
+
+end Retract
+
+/--
+```
+  X -------> Z -------> X
+  |          |          |
+  f          g          f
+  |          |          |
+  v          v          v
+  Y -------> W -------> Y
+
+```
+A morphism `f : X ⟶ Y` is a retract of `g : Z ⟶ W` if there are morphisms `i : f ⟶ g`
+and `r : g ⟶ f` in the arrow category such that `i ≫ r = 𝟙 f`. -/
+abbrev RetractArrow {X Y Z W : C} (f : X ⟶ Y) (g : Z ⟶ W) := Retract (Arrow.mk f) (Arrow.mk g)
+
+namespace RetractArrow
+
+variable {X Y Z W : C} {f : X ⟶ Y} {g : Z ⟶ W} (h : RetractArrow f g)
```

Comment:
You may add that if `RetractArrow f g`, then `Retract X Z` (this should be named `RetractArrow.left`), and `Retract Y W`.